### PR TITLE
Migrating a bunch of Ara related code to GEM Tables

### PR DESCRIPTION
**Pull request by dleblanc**
# Citrine Python PR

This does a large rename of classes inside citrine.ara to citrine.gemtables, and makes related changes to classnames and documentation to convert to the new nomenclature (See PLA-4258 for details).

Though the acronym is GEM, in class names this will be eg: GemSomething.

This PR does not include changes related to backend JSON endpoint and field name changes, those will be handled separately.

With the exception of a couple error messages (which I will update in the e2e tests to handle both cases), all of these changes should be backwards compatible with the shadowed/aliased names provided. Deprecations are provided where possible.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [x] I have added tests for 100% coverage
- [x] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.
- [x] I have bumped the version in setup.py

### CODE REVIEW

File: docs/source/workflows/data_sources.rst [hunk 1]
```diff
@@ -143,18 +143,18 @@ A CSV data source and :class:`~citrine.informatics.predictors.IngredientsToSimpl
 Ara Table Data Source
```
Comment (kroenlein) [inline docs/source/workflows/data_sources.rst:143]:
```suggestion
GEM Table Data Source
```

File: src/citrine/_utils/functions.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
+import inspect
```
Comment (djack) [inline src/citrine/_utils/functions.py:1]: huh. TIL

File: tests/_util/source_mod.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+class ExampleClass():
```
Comment (djack) [inline tests/_util/source_mod.py:1]: Is this here deliberately?

Comment (djack) [reply tests/_util/source_mod.py:1]: Ah, nevermind. Used in the test. Got it.

File: tests/resources/test_table_config.py [hunk 4]
```diff
@@ -69,20 +70,35 @@ def test_get_ara_definition_metadata(collection, session):
...
+    table_config = TableConfig(name="foo", description="bar", rows=[], columns=[], variables=[], datasets=[])
+    assert table_config.config_uid is None
+    assert table_config.version_number is None
+
+
+def test_init_table_config_with_old_definition_uid():
+    uid = UUID('6b608f78-e341-422c-8076-35adc8828566')
+    table_config = TableConfig(name="foo", description="bar", rows=[], columns=[], variables=[], datasets=[],
+                               definition_uid=uid)
+    assert table_config.config_uid == uid
```
Comment (djack) [inline tests/resources/test_table_config.py:87]: :+1:

**Comment by dleblanc** [comment]
> There appear to be some knock-on effects with these changes and nextgen-devkit, e.g. some library imports that were sneaking in through `import *`.

What do you mean by this @kroenlein ?

File: docs/source/workflows/data_sources.rst [hunk 5]
```diff
@@ -143,18 +143,18 @@ A CSV data source and :class:`~citrine.informatics.predictors.IngredientsToSimpl
...
 ---------------------
 
-An :class:`~citrine.informatics.data_sources.AraTableDataSource` references an Ara table.
-As explained more in the :doc:`documentation <../data_extraction>`, Ara tables provide a structured version of on-platform data.
+An :class:`~citrine.informatics.data_sources.GemTableDataSource` references a GEM Table.
+As explained more in the :doc:`documentation <../data_extraction>`, GEM Tables provide a structured version of on-platform data.
 Ara tables are specified by the display table uuid and version number.
```
[the remark this answers is not in the document]
Comment (dleblanc) [reply docs/source/workflows/data_sources.rst]: Done

File: docs/source/workflows/data_sources.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (dleblanc) [reply docs/source/workflows/data_sources.rst]: Done

File: docs/source/data_extraction.rst [hunk 6]
```diff
@@ -103,24 +103,24 @@ If the process template's allowed names includes, for example, "flour" then ther
 Previewing tables
 -----------------
 
-Calling :func:`~citrine.resources.project.Project.ara_definitions` on a project returns an :class:`~citrine.resources.ara_definition.AraDefinitionCollection` object, which facilitates access to the collection of all Ara definitions visible to a Project.
-Via such an object, one can preview a draft AraDefinition on an explicit set of Material Histories, defined by their root materials:
+Calling :func:`~citrine.resources.project.Project.table_configs` on a project returns an :class:`~citrine.resources.ara_definition.TableConfigCollection` object, which facilitates access to the collection of all Ara definitions visible to a Project.
```
[the remark this answers is not in the document]
Comment (dleblanc) [reply docs/source/data_extraction.rst]: Done

**Comment by kroenlein** [comment]
> > There appear to be some knock-on effects with these changes and nextgen-devkit, e.g. some library imports that were sneaking in through `import *`.
> 
> What do you mean by this @kroenlein ?

For example, nextgen-devkit:end-to-end-tests/tests/preload/cake.py is accidentally importing typing.Tuple through `from citrine.ara.variables import *`.  I'm in the midst of some import cleanups to address this, but please don't merge until citrine-python  passes integration tests.

**Pull request merged**